persistence/docs/SAVEPOINTS_ROLLBACKS.md

```diff
@@ -0,0 +1,146 @@
+- To allow nodes to recover from an unforeseen crash (bugs)
+- To facilitate socially coordinated rollbacks of the chain to a specific block height in the case of a consensus failure/chain-halt
+- To improve operator experience when managing and upgrading their fleet of nodes
+- And more...
+
+Performing operations like: copying datadirs, compressing them, etc. is probably not the best approach.
+Having a first-class support for savepoints and rollbacks would be, IMHO, a much better solution.
+
+A more advanced/flexible version of **Savepoints** could be to efficiently serialize the state into an artifact (non-trivial also because we have multiple data-stores but definitely possible), maybe leveraging previous savepoints to only store the changes that have been made since the last one and/or using some version of a WAL (Write-Ahead log that records the changes that happened).
+
+#### Savepoints
+
+A savepoint must have the following properties:
+
+- It must be able to be created at any block height
+- It must be self-contained and/or easy to move around (e.g. a file/archive)
+- The action of creating a savepoint must be atomic (i.e. it must be impossible to create a savepoint that is incomplete)
+- The action of creating a savepoint must be easy to perform (e.g. a CLI command and/or a flag being passed to the node binary when starting it)
+- The action of creating a savepoint must be as fast as possible
+- The operator should be informed with meaningful messages about the progress of the savepoint creation process (telemetry, logging and stdout come to mind)
+- It should be, as much as possible, compact
+- It must have some form of integrity check mechanism (e.g. checksum/hash verification and maybe even a signature that could be very useful in the case of a social rollback)
+
+#### Rollbacks
+
+The following are some of the properties that a rollback mechanism must have:
+
+- If the operator requested a rollback, regardless of the internal state of the node, it must be able to rollback to the requested block height, provided a valid savepoint
+- The rollback process must be atomic (i.e. it must be impossible to rollback to a block height has incomplete/invalid state)
+- The rollback process must be easy to perform (e.g. a CLI command and/or a flag being passed to the node binary when starting it)
+- The rollback process must be as fast as possible
+- The operator should be informed in a meaningful way about the progress of the rollback process (telemetry, logging and stdout come to mind)
+
+
+### Further improvements
+
+- Savepoints could be disseminated/retrieved using other networks like Tor/IPFS/etc., this would free-up bandwidth on the main network and could be used in conjunction with `FastSync` to speed up the process of bootstrapping a new node without overwhelming the `P2P` network with a lot of traffic that is not `Protocol`-related. This could be very important when the network reaches critical mass in terms of scale. Hopefully soon.
```

Comment:
I see, you really do look at savepoints as snapshots. 

Do you see the solution for a node-specific savepoint (i.e. I updated one tree but not another and crashed) and a network wide (i.e. a data dir snapshot) as having the same solution?

Reply:
I think it can very much build on top of what I have in mind.

Similarly to what databases do, we'd need some `data integrity check` at startup that verifies that all the datastores are correct.

It can very much rely on the statehash I guess...

How?

KISS, back of the envelope implementation:
- retrieve the latest valid statehash (from file, from `SuperValidators`, from P2P, TBD)
- at startup check if it matches the current, computed state hash
- if not, rollback to latest savepoint and sync from there

Reply:
> retrieve the latest valid statehash (from file, from SuperValidators, from P2P, TBD)

This is what "weak subjectivity" is. From [Vitalik's 2014 article](https://blog.ethereum.org/2014/11/25/proof-stake-learned-love-weak-subjectivity)

> Weak subjectivity is exactly the correct solution. It solves the long-range problems with proof of stake by relying on human-driven social information, but leaves to a consensus algorithm the role of increasing the speed of consensus from many weeks to twelve seconds and of allowing the use of highly complex rulesets and a large state. T

--- 

> if not, rollback to latest savepoint and sync from there
> at startup check if it matches the current, computed state hash

I think there are implementation nuances here, but I think you've got it.

--- 

I think it's worth adding (in the moonshot section) that we can potentially use Mina to verify the snapshot hash using only the latest state and the `genesis.json` file

https://minaprotocol.com/lightweight-blockchain

<img width="931" alt="Screenshot 2023-03-07 at 5 22 17 PM" src="https://user-images.githubusercontent.com/1892194/223594923-ffb42403-fafc-4c67-ac02-09e9e839db6a.png">

Reply:
> I think there are implementation nuances here, but I think you've got it.

Music to my ears. 

I am going to read that article tonight. Thanks for sharing. 🙏

> I think it's worth adding (in the moonshot section) that we can potentially use Mina to verify the snapshot hash using only the latest state and the genesis.json file

Added